libs/utils/src/celix_err.c

```diff
@@ -0,0 +1,166 @@
+        err->pos = 0; //no entry
+        celix_status_t status = celix_tss_set(celix_err_tssKey, err);
+        if (status != CELIX_SUCCESS) {
+            fprintf(stderr, "Failed to set thread specific storage for celix_err\n");
+            free(err);
+            err = NULL;
+        }
+    } else {
+        fprintf(stderr, "Failed to allocate memory for celix_err\n");
+    }
+    return err;
+}
+
+__attribute__((constructor)) void celix_err_initThreadSpecificStorageKey() {
+    celix_status_t status = celix_tss_create(&celix_err_tssKey, celix_err_destroyTssErr);
+    if (status == CELIX_SUCCESS) {
+        celix_err_tssKeyInitialized = true;
+    } else {
+        fprintf(stderr,"Failed to create thread specific storage key for celix_err\n");
+    }
+}
+
+__attribute__((destructor)) void celix_err_deinitThreadSpecificStorageKey() {
+    if (!celix_err_tssKeyInitialized) {
+        fprintf(stderr, "celix_err_tssKey is not initialized\n");
+        return;
+    }
+
+    celix_status_t status = celix_tss_delete(celix_err_tssKey);
+    if (status != CELIX_SUCCESS) {
+        fprintf(stderr,"Failed to delete thread specific storage key for celix_err\n");
+    }
+}
+
+#endif
+
+const char* celix_err_popLastError() {
+    const char* result = NULL;
+    celix_err_t* err = celix_err_getTssErr();
+    if (err && err->pos > 0) {
+        //move back to start last error message
+        err->pos -= 1; //move before \0 in last error message
+        while (err->pos > 0 && err->buffer[err->pos-1] != '\0') {
+            err->pos -= 1;
+        }
+        result = err->buffer + err->pos;
+    }
+    return result;
+}
+
+int celix_err_getErrorCount() {
+    int result = 0;
+    celix_err_t* err = celix_err_getTssErr();
+    for (int i = 0; err && i < err->pos; ++i) {
+        if (err->buffer[i] == '\0') {
+            result += 1;
+        }
+    }
+    return result;
+}
+
+void celix_err_resetErrors() {
+    celix_err_t* err = celix_err_getTssErr();
+    if (err) {
+        err->pos = 0; //no entry
+    }
+}
+
+void celix_err_push(const char* msg) {
+    celix_err_t* err = celix_err_getTssErr();
+    if (err) {
+        size_t len = strnlen(msg, CELIX_ERR_BUFFER_SIZE);
+        if (err->pos + len + 1 <= CELIX_ERR_BUFFER_SIZE) {
+            strcpy(err->buffer + err->pos, msg);
+            err->pos += len + 1;
+        } else {
+            fprintf(stderr, "Failed to add error message '%s' to celix_err\n", msg);
+        }
+    }
+}
+
+void celix_err_pushf(const char* format, ...) {
+    va_list args;
+    va_start(args, format);
+    celix_err_t* err = celix_err_getTssErr();
+    if (err) {
+        size_t len = vsnprintf(err->buffer + err->pos, CELIX_ERR_BUFFER_SIZE - err->pos, format, args);
+        if (err->pos + len + 1 <= CELIX_ERR_BUFFER_SIZE) {
+            err->pos += len;
+            err->buffer[err->pos++] = '\0';
+        } else {
+            fprintf(stderr, "Failed to add error message '");
+            vfprintf(stderr, format, args);
```

Comment:
`args` may have been modified by `vsnprintf`. 
Therefore we need `va_copy` before `vsnprintf`.